### PR TITLE
Job ban 2.0: API's Revenge

### PR DESCRIPTION
**Pull request by francinum**
[CLEAN]
## About the PR 

Reworks the job ban system to rely on the API over a save file.

* Bans now cache at the player level.
    * This cache is invalidated on logout.
* Ban panel is slightly less nasty
    * It can now be refreshed
    * jobban-panel now takes a ckey to bring up an arbitrary ckey's jobban UI.
## Why's this needed? 

...Have you seen the current implementation?

### CODE REVIEW

File: code/modules/admin/admin_verbs.dm [hunk 1]
```diff
@@ -657,20 +656,6 @@ var/list/special_pa_observing_verbs = list(
 		src.holder.s_respawn()
 	return
 
-/client/proc/jobbans()
```
Comment (ZeWaka) [inline code/modules/admin/admin_verbs.dm:660]: If you're deleting the interface, how are admins supposed to view current jobbans for a ckey?

Comment (francinum) [reply code/modules/admin/admin_verbs.dm:660]: Player options, for offline players I'm just about to add a replacement for this panel to view them.

File: code/modules/admin/admin.dm [hunk 2]
```diff
@@ -486,6 +486,20 @@ var/global/noir = 0
 				alert("You need to be at least a Secondary Administrator to remove bans.")
 		/////////////////////////////////////end ban stuff
 
+		if("jobb_unban_disconnected")
+			if (src.level >= LEVEL_SA)
+				var/t = href_list["target"]
+				var/r = href_list["type"]
```
Comment (ZeWaka) [inline code/modules/admin/admin.dm:492]: poor variable names

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/modules/admin/admin.dm (hunk 2 above)
Comment (francinum) [reply code/modules/admin/admin.dm]: Entire segment reworked.

File: code/world.dm [hunk 3]
```diff
@@ -504,7 +483,7 @@ var/f_color_selector_handler/F_Color_Selector
 		if (config.server_name != null && config.server_suffix && world.port > 0)
 			config.server_name += " #[serverKey]"
 
-		precache_unban_txt() //Wire: left in for now because jobbans still use the shitty system
+		//precache_unban_txt() //Wire: left in for now because jobbans still use the shitty system //Francinum: Merci~
```
Comment (ZeWaka) [inline code/world.dm:486]:
```suggestion
```

**Review by ZeWaka — APPROVED** [review]

**Review by ZeWaka — APPROVED** [review]
DNM due to api-side modifications on Wire's end. Wire can remove this DNM.

**Review by Wirewraith — CHANGES_REQUESTED** [review]
I have made the API-side changes (sorry for the delay).

Ready to merge pending changes from review.

File: code/modules/admin/jobban.dm [hunk 4]
```diff
@@ -1,82 +1,90 @@
...
+		if ("Main")
+			server = "main"
+		if ("Roleplay Overflow")
+			server = "main2"
+		if ("Main Overflow")
+			server = "main3"
+	if(apiHandler.queryAPI("jobbans/add", list("ckey"=M,"rank"=rank, "akey"=akey, "applicable_server"=server)))
+		var/datum/player/player = make_player(M) //Recache the player.
+		if(player)
+			player.cached_jobbans = apiHandler.queryAPI("jobbans/get/player", list("ckey"=MACHINES_ANNOUNCEMENTS), 1)[M]
```
[the remark this answers is not in the document]
Comment (francinum) [reply code/modules/admin/jobban.dm:22]: ...Where the hell did that come from?

Comment (francinum) [inline code/modules/admin/jobban.dm:22]:
```suggestion
			player.cached_jobbans = apiHandler.queryAPI("jobbans/get/player", list("ckey"=M), 1)[M]
```

**Comment by ZeWaka** [comment]
Wire gave approval via Discord.

**Pull request merged**